### ISSUE TITLE
getAutoUserns() returns non consecutive usern-ns ID mappings

### ISSUE THREAD

**Issue by ctalledo** [containers/storage#852]
I am seeing that getAutoUserns() (in store.go) returns non consecutive user-ns ID mappings in scenarios where it should have returned a continuous mapping.

For example, when I call this function to request allocation of 65536 UIDs from host ID range starting at 296608, where mapping {ContainerID: 0, HostID: 296608, Size:1} is passed in as a "used UID", the function returns the following mappings:

```
{ContainerID:0 HostID:296608 Size:1}
{ContainerID:1 HostID:296610 Size:65535} 
```

Note how host ID 296609 was skipped, even though no other container is mapped to that host ID.  Ideally the second range would have started at HostID 296609.

I looked at the code, and it seems function `findAvailableRange()` is the real culprit. It's not returning the correct range in some cases. 

For example, adding this simple sub-test (resembling the example above) to unit test TestFindAvailableIDRange() triggers the problem:

```golang
   avail = []idtools.IDMap{                                                                                                                                                                                                                                                                                                   
      {                                                                                                                                                                                                                                                                                                                       
         ContainerID: 0,                                                                                                                                                                                                                                                                                                      
         HostID:      296608,                                                                                                                                                                                                                                                                                                 
         Size:        65536,                                                                                                                                                                                                                                                                                                  
      },                                                                                                                                                                                                                                                                                                                      
   }                                                                                                                                                                                                                                                                                                                          
                                                                                                                                                                                                                                                                                                                              
   used = []idtools.IDMap{                                                                                                                                                                                                                                                                                                    
      {                                                                                                                                                                                                                                                                                                                       
         ContainerID: 0,                                                                                                                                                                                                                                                                                                      
         HostID:      296608,                                                                                                                                                                                                                                                                                                 
         Size:        1,                                                                                                                                                                                                                                                                                                      
      },                                                                                                                                                                                                                                                                                                                      
   }                                                                                                                                                                                                                                                                                                                          
   ret, err = findAvailableIDRange(65535, avail, used)                                                                                                                                                                                                                                                                        
```

This returns the range `{0 296609 65535}`, instead of range `{1 296609 65535}` as expected (i.e., the ContainerID is incorrect).

As a result, availableUIDs in line 416 below has the incorrect ContainerID, and this in turn causes the call to `subtractAll()` in line 424 to bump up the host ID to 296610, thereby incorrectly skipping 296609.

```
 416 │    availableUIDs, availableGIDs, err = findAvailableRange(size, size, availableUIDs, availableGIDs, usedUIDs, usedGIDs)                                                                                                                                                                                                
 417 │    if err != nil {                                                                                                                                                                                                                                                                                                     
 418 │       return nil, nil, err                                                                                                                                                                                                                                                                                             
 419 │    }                                                                                                                                                                                                                                                                                                                   
 420 │                                                                                                                                                                                                                                                                                                                        
 421 │    // We need to make sure the specified container IDs are also dropped from the automatic                                                                                                                                                                                                                             
 422 │    // namespaces we have found.                                                                                                                                                                                                                                                                                        
 423 │    if len(options.AdditionalUIDMappings) > 0 {                                                                                                                                                                                                                                                                         
 424 │       availableUIDs = subtractAll(availableUIDs, options.AdditionalUIDMappings, false)                                                                                                                                                                                                                                 
 425 │    }  
```

**Comment by giuseppe** [comment]
Hi Cesar, thanks for the detailed report!

If I run locally the test case you've reported:

```
func TestConsecutiveMapping(t *testing.T) {
	avail := []idtools.IDMap{
		{
			ContainerID: 0,
			HostID:      296608,
			Size:        65536,
		},
	}
	used := []idtools.IDMap{
		{
			ContainerID: 0,
			HostID:      296608,
			Size:        1,
		},
	}
	ret, err := findAvailableIDRange(65535, avail, used)
	assert.Nil(t, err)
	fmt.Printf("%v\n", ret)
}
```

I get: `[{0 296609 65535}]`

Do you see a different result?

The ContainerID always starts at 0, it is incremented sequentially. e.g. with:

```
	avail := []idtools.IDMap{
		{
			ContainerID: 0,
			HostID:      296608,
			Size:        65538,
		},
	}
	used := []idtools.IDMap{
		{
			ContainerID: 0,
			HostID:      296608,
			Size:        1,
		},
		{
			ContainerID: 0,
			HostID:      296612,
			Size:        1,
		},
	}
```

I get: `[{0 296609 3} {3 296613 65532}]`

`ContainerID[i] = ContainerID[i-1] + Size[i-1]`, with `ContainerID[0] = 0`.

So the ContainerID for the first range is always 0, for the second range is the size of the first range and so on.

**Comment by ctalledo** [comment]
HI @giuseppe , thanks for the quick response.

> I get: [{0 296609 65535}]

Yes, that's the problem. Function findAvailableIDRange() should have returned `[{1 296609 65535}]` instead, because container ID 0 (mapped to host ID 296608) is already "used" (i.e., it's not available).  

As a result of findAvailableIDRange() returning that incorrect container ID range, when the calling function (`getAutoUserNS()`) calls `subtractAll()` in L424 and passes in the invalid range ({0 296609 65535}), the subtraction notices the collision on container ID 0, so bumps up the range to `1 296610 65535`, thus skipping 296609 inadvertendly.

For example, if I instrument function `getAutoUserNS()` (userns.go) and add the logging:

```
 412 │    // Make sure the specified additional IDs are not used as part of the automatic                                                                                                                                                                                                                                     
 413 │    // mapping                                                                                                                                                                                                                                                                                                          
 414 │    usedUIDs = append(usedUIDs, options.AdditionalUIDMappings...)                                                                                                                                                                                                                                                       
 415 │    usedGIDs = append(usedGIDs, options.AdditionalGIDMappings...)                                                                                                                                                                                                                                                       
 416 │                                                                                                                                                                                                                                                                                                                        
 417 │    logrus.Infof("(1) availableUIDs = %v", availableUIDs)                                                                                                                                                                                                                                                               
 418 │                                                                                                                                                                                                                                                                                                                        
 419 │    availableUIDs, availableGIDs, err = findAvailableRange(size, size, availableUIDs, availableGIDs, usedUIDs, usedGIDs)                                                                                                                                                                                                
 420 │    if err != nil {                                                                                                                                                                                                                                                                                                     
 421 │       return nil, nil, err                                                                                                                                                                                                                                                                                             
 422 │    }                                                                                                                                                                                                                                                                                                                   
 423 │                                                                                                                                                                                                                                                                                                                        
 424 │    logrus.Infof("(2) availableUIDs = %v", availableUIDs)                                                                                                                                                                                                                                                               
 425 │                                                                                                                                                                                                                                                                                                                        
 426 │    // We need to make sure the specified container IDs are also dropped from the automatic                                                                                                                                                                                                                             
 427 │    // namespaces we have found.                                                                                                                                                                                                                                                                                        
 428 │    if len(options.AdditionalUIDMappings) > 0 {                                                                                                                                                                                                                                                                         
 429 │       availableUIDs = subtractAll(availableUIDs, options.AdditionalUIDMappings, false)                                                                                                                                                                                                                                 
 430 │    }                                                                                                                                                                                                                                                                                                                   
 431 │    if len(options.AdditionalGIDMappings) > 0 {                                                                                                                                                                                                                                                                         
 432 │       availableGIDs = subtractAll(availableGIDs, options.AdditionalGIDMappings, false)                                                                                                                                                                                                                                 
 433 │    }                                                                                                                                                                                                                                                                                                                   
 434 │                                                                                                                                                                                                                                                                                                                        
 435 │    logrus.Infof("(3) availableUIDs = %v", availableUIDs)         
```

This is the result:

```
level=info msg="(1) availableUIDs = [{0 296608 655360}]"                                                                                                                                                                                          
level=info msg="(2) availableUIDs = [{0 296609 65536}]"                                                                                                                                                                                           
level=info msg="(3) availableUIDs = [{1 296610 65535}]"   
```

Notice how 296609 got bumped up to 296610. I think the second line should have returned `{1 296609 65535}` since that's the available range (because `0 296608 1` is used already).

**Comment by giuseppe** [comment]
thanks for the additional info.

Could you also print the content of `AdditionalUIDMappings`?

I think it could be an issue in `subtractAll`.

@likan999 could you please take a look?  Do you think it is something that `subtractAll` should take into account?

**Comment by ctalledo** [comment]
> Could you also print the content of `AdditionalUIDMappings`?

Sure:

```
Mar 11 21:30:05 k8s-node crio[69234]: time="2021-03-11 21:30:05.764873499Z" level=info msg="(1) availableUIDs = [{0 296608 655360}], additionalUIDMappings = [{0 296608 1}]"                                                                                                                                             
Mar 11 21:30:05 k8s-node crio[69234]: time="2021-03-11 21:30:05.764926986Z" level=info msg="(2) availableUIDs = [{0 296609 65536}], additionalUIDMappings = [{0 296608 1}]"                                                                                                                       
Mar 11 21:30:05 k8s-node crio[69234]: time="2021-03-11 21:30:05.764954630Z" level=info msg="(3) availableUIDs = [{1 296610 65535}], additionalUIDMappings = [{0 296608 1}]"  
```

Notice also how in the second line the availableUIDs slice has an invalid size = 65536; we expect 65535 since one of the mappings is already used.
 
> I think it could be an issue in `subtractAll`.
> 
> @likan999 could you please take a look? Do you think it is something that `subtractAll` should take into account?

I took a closer look, and I *think* the bug is in findAvailableIDRange(), as marked below:

```
func findAvailableIDRange(size uint32, availableIDs, usedIDs []idtools.IDMap) ([]idtools.IDMap, error) {
	var avail []idtools.IDMap

	// ContainerID will be adjusted later.
	for _, i := range availableIDs {
		n := idtools.IDMap{
			ContainerID: 0,
			HostID:      i.HostID,
			Size:        i.Size,
		}
		avail = append(avail, n)
	}
	avail = subtractAll(avail, usedIDs, true)

	currentID := 0                    <<<< BUG HERE: should be "currentID := avail[0].ContainerID"
	remaining := size
	// We know the size for each intervals, let's adjust the ContainerID for each
	// of them.
	for i := 0; i < len(avail); i++ {
		avail[i].ContainerID = currentID
```

The `currentID := 0` line assumes the containerID range starts at zero, but misses the fact that this containerID may not be available (because is part of the userIDs array).

In addition, in function `getAutoUserNS()`, when calling `findAvailableRange()` , the `size` parameter is incorrect in my view:

```
availableUIDs, availableGIDs, err = findAvailableRange(size, size, availableUIDs, availableGIDs, usedUIDs, usedGIDs)
```

It's currently passing in the size of the entire range requested, including the usedUIDs. It seems to me it should pass the size of the range requested *minus* the used UIDs (because that's the number of available UIDs we are looking for).

I tried both of these changes and I can see things work better now:


```
Mar 11 21:46:06 k8s-node crio[70387]: time="2021-03-11 21:46:06.422211738Z" level=error msg="(1) availableUIDs = [{0 296608 655360}], additionalUIDMappings = [{0 296608 1}]"                                                                                                                                                 
Mar 11 21:46:06 k8s-node crio[70387]: time="2021-03-11 21:46:06.422278670Z" level=error msg="(2) availableUIDs = [{1 296609 65535}], additionalUIDMappings = [{0 296608 1}]"                                                                                                                                                  
Mar 11 21:46:06 k8s-node crio[70387]: time="2021-03-11 21:46:06.422327022Z" level=error msg="(3) availableUIDs = [{1 296609 65535}], additionalUIDMappings = [{0 296608 1}]"
```

Notice how the containerID, hostIDs, and size look correct now.

**Comment by giuseppe** [comment]
thanks for your detailed analysis.

Could you open a PR with your changes and the test case you've written?

**Comment by ctalledo** [comment]
> Could you open a PR with your changes and the test case you've written?

Sure. Apart from the unit tests in userns_test.go, are there any integration tests (say in the CRI-O repo) that I could run to make sure all is good?

**Comment by ctalledo** [comment]
FYI: PR is [here](https://github.com/containers/storage/pull/853).

**Comment by ctalledo** [comment]
Related to this issue, when using `crictl` + CRI-O to play around with the pod user-ns ID mappings, I am seeing some strange behavior. 

The following are observations without the proposed fix for this issue (i.e., with upstream containers/storage). My `/etc/subuid` file has `containers:296608:655360`.

#### 1) Pod with `io.kubernetes.cri-o.userns-mode": "auto:size=65536"`  and `run_as_user: 0`  

CRIO generates the following userns UID mappings: 

{ContainerID:0 HostID:0 Size:1}
{ContainerID:1 HostID:296609 Size:65535}             <<< OK but not ideal as 296608 is unused.

#### 2) Pod with `io.kubernetes.cri-o.userns-mode": "auto:size=65536"`  and `run_as_user: 296608`

{ContainerID:0 HostID:296608 Size:1}
{ContainerID:1 HostID:296610 Size:65535}              <<< BUG: should be 296609 (issue tracked by this PR)

#### 3) Pod with `io.kubernetes.cri-o.userns-mode": "auto:size=65536"`  and `run_as_user: 296708`

(i.e., I want the pod to run with user 1000 inside the pod, which maps to host ID 296708).

{ContainerID:0 HostID:296708 Size:1}              
{ContainerID:1 HostID:296609 Size:99}             
{ContainerID:100 HostID:296709 Size:65436} 

Looks very strange; I expected a single mapping such as {ContainerID: 0 HostID: 296608: Size: 65536}, with the OCI spec indicating to run as user 1000 inside the container (i.e., process.user.uid = "1000"). This way, user 1000 in the container is mapped to host ID 297608 as requested. 

#### 4) Pod with `"io.kubernetes.cri-o.userns-mode": "private:uidmapping=0:296608:65536;gidmapping=0:296608:65536"` and `run_as_user: 296608`

CRI-O complains: `cannot use uidmapping or gidmapping if not running as root`. That's strange, I am trying to map host ID 296608 to container ID 0. How else should I do this?

#### 5) Pod with `"io.kubernetes.cri-o.userns-mode": "private:uidmapping=0:296608:65536;gidmapping=0:296608:65536"` and `run_as_user: 0`

{ContainerID:0 HostID:296608 Size:65536} 
{ContainerID:65536 HostID:0 Size:1}                     <<< BUG: invalid mapping


This behavior seems broken, unless I am missing something basic here.

**Comment by giuseppe** [comment]
> #### 4) Pod with `"io.kubernetes.cri-o.userns-mode": "private:uidmapping=0:296608:65536;gidmapping=0:296608:65536"` and `run_as_user: 296608`
> 
> CRI-O complains: `cannot use uidmapping or gidmapping if not running as root`. That's strange, I am trying to map host ID 296608 to container ID 0. How else should I do this?

Managing directly the IDs is a privileged operation.  You are not allowed to use it unless the POD is running as root (`run_as_user: 0`).

> #### 5) Pod with `"io.kubernetes.cri-o.userns-mode": "private:uidmapping=0:296608:65536;gidmapping=0:296608:65536"` and `run_as_user: 0`
> 
> {ContainerID:0 HostID:296608 Size:65536}
> {ContainerID:65536 HostID:0 Size:1} <<< BUG: invalid mapping
> 
> This behavior seems broken, unless I am missing something basic here.

`run_as_user: 0` refers to root on the host, because that is what the rest of Kubernetes knows, so it is mapped to the first available ID inside the user namespace.

Would you like to run as root inside the container but mapped to a different user on the host?

**Comment by ctalledo** [comment]
> Managing directly the IDs is a privileged operation. You are not allowed to use it unless the POD is running as root (run_as_user: 0).

Not sure I follow. The pod is the entity whose UIDs are going to be mapped. The entity doing the mapping is K8s + CRIO. The latter run as root and thus can do the mapping per the pod's spec. 

> Would you like to run as root inside the container but mapped to a different user on the host?

Yes exactly: I want to have a pod with container UIDs 0->65535  mapped to host IDs 296608-> (296608 + 65535). This way root in the pod is mapped to a non-root user in the host, thereby providing isolation promised by user-namespaces. 

I managed to get crictl + CRIO to do that by specifying `io.kubernetes.cri-o.userns-mode": "auto:size=65536"` and omitting the "run_as_user".  That's good, but not ideal, since I have no mechanism to run as user 1000 inside the pod, where the pod container IDs are mapped say to host IDs 296608 -> (296608 + 65535) and user 1000 is thus mapped to 296708.

**Comment by giuseppe** [comment]
> > Managing directly the IDs is a privileged operation. You are not allowed to use it unless the POD is running as root (run_as_user: 0).
> 
> Not sure I follow. The pod is the entity whose UIDs are going to be mapped. The entity doing the mapping is K8s + CRIO. The latter run as root and thus can do the mapping per the pod's spec.

technically yes, there is nothing preventing CRI-O to set that mapping.  The problem though is that there are a lot of higher level policies (e.g.,  OpenShift SCCs) that can prevent a POD to run as root.  To avoid circumventing these policies, we limit the usage of directly setting the mappings unless the POD is already running as root, so it means it is privileged enough that the mappings  won't grant access to other IDs that it cannot already access. 

> > Would you like to run as root inside the container but mapped to a different user on the host?
> 
> Yes exactly: I want to have a pod with container UIDs 0->65535 mapped to host IDs 296608-> (296608 + 65535). This way root in the pod is mapped to a non-root user in the host, thereby providing isolation promised by user-namespaces.
> 
> I managed to get crictl + CRIO to do that by specifying `io.kubernetes.cri-o.userns-mode": "auto:size=65536"` and omitting the "run_as_user". That's good, but not ideal, since I have no mechanism to run as user 1000 inside the pod, where the pod container IDs are mapped say to host IDs 296608 -> (296608 + 65535) and user 1000 is thus mapped to 296708.

Have you tried using the `keep-id` option?

"io.kubernetes.cri-o.userns-mode": "auto:keep-id=true"

**Comment by likan999** [comment]
I have a simpler PR #856 that passes the unit tests in #853. Please let me know if the PR works. Thanks.

**Comment by ctalledo** [comment]
> Yes exactly: I want to have a pod with container UIDs 0->65535 mapped to host IDs 296608-> (296608 + 65535). This way root in the pod is mapped to a non-root user in the host, thereby providing isolation promised by user-namespaces.
>
> I managed to get crictl + CRIO to do that by specifying io.kubernetes.cri-o.userns-mode": "auto:size=65536" and omitting the "run_as_user".

I just realized that only works with CRI-O 1.20, but not with top-of-tree master. With the latter, omitting "run_as_user" is equivalent to "run_as_user = 0", so CRI-O generates the following mapping:

```
{ContainerID:0 HostID:0 Size:1}
{ContainerID:1 HostID:296609 Size:65535}
``` 

This is not what I want since it maps root in the container to root in the host; I want ContainerID 0 -> HostID 296608.

> Have you tried using the keep-id option?
>
> "io.kubernetes.cri-o.userns-mode": "auto:keep-id=true"

Just tried it, but it still does not do what's needed:

a) "io.kubernetes.cri-o.userns-mode": "auto:size=65536;keep-id=true" :

{ContainerID:0 HostID:0 Size:1}                             <<< Not what we want
{ContainerID:1 HostID:296609 Size:65535} 

b) Same as (a), but with "run_as_user = 296608":

{ContainerID:0 HostID:296609 Size:65536}            <<< Not what we want
{ContainerID:296608 HostID:296608 Size:1}          <<< Not what we want

**Comment by likan999** [comment]
After some closer look, it seems my previous mindset is incorrect. There is not any invariant to maintain the difference between host and container ids (sorry @giuseppe). What it really should happen is
1) read availableHostIds from /etc/subuids;
2) availableHostIds := subtract(availableHostIds, hostIdsOf(usedIdsMap)+hostIdsOf(additionalIdsMap));
3) compute requestedSize as currently is doing (not subtracting any used ids as #853); the containerIds range is [0, requestedSize);
4) containerIds := subtract(containerIds, containerIdsOf(additionalIdsMap));
5) let n be the size of containerIds; take n ids from availableHostIds;
6) now form a mapping from the n ids in step 5 to containerIds; together with the additionalIdsMap is the mapping to return.

The additionalIdsMap contains the mapping from host uid => container id 0. By step 4, we will now only allocate requestSize-1 instead of requestSize ids from the available container ids, which is the current problem in #856.

If the above steps sounds reasonable to you @giuseppe @ctalledo , I'll implement this in #856.
The above steps assume we always map to the [0, n) in container ids range. If we ever need to support other container id ranges, #856 would provide arithmetics to manipulate id ranges thus should make it straightforward.

**Comment by giuseppe** [comment]
@likan999 thanks for taking a look at this.

Yes, these steps look reasonable to me

**Comment by ctalledo** [comment]
Sounds good @likan999  ... thanks for taking a closer look. I wonder if we need better unit tests too, because the existing tests where not catching some of the problem in the original fix in #853 and in the alternative fix in #856.

**Comment by ctalledo** [comment]
Hi @giuseppe ,

Back on this topic:

> I want to have a pod with container UIDs 0->65535 mapped to host IDs 296608-> (296608 + 65535). This way root in the pod is mapped to a non-root user in the host, thereby providing isolation promised by user-namespaces.

As I mentioned, I managed to get crictl + CRI-O to do that by specifying io.kubernetes.cri-o.userns-mode": "auto:size=65536" and omitting the "run_as_user".  But this only works on CRI-O release 1.20.  I can't find a way to do this on CRI-O top-of-tree (i.e., omiting "run_as_user" is equivalent to run_as_user = 0, which maps UID 0 in the pod to UID 0 in the host, breaking user-ns isolation).

I tried all the user-ns mode annotations such as "keep-id=true", "private", etc., but with none of them I can get this to work.

How do I do this?

**Comment by giuseppe** [comment]
@haircommander FYI.

@ctalledo that is right, without an explicit `run_as_user` then Kubernetes assumes it is root.  You'd need to manually set a `run_as_user` so that the rest of Kubernetes knows what is the user that runs the container

**Comment by ctalledo** [comment]
> @ctalledo that is right, without an explicit run_as_user then Kubernetes assumes it is root. You'd need to manually set a run_as_user so that the rest of Kubernetes knows what is the user that runs the container

Thanks @giuseppe; but how can I set "run_as_user" to a host ID if I don't know what host ID to set it to, given that CRI-O is automatically choosing the host ID via the `"io.kubernetes.cri-o.userns-mode": "auto:size=65536"` annotation? 

In other words, if I use "auto:size=65536" (which I like because CRI-O chooses the UID mapping automatically), I don't know what to set "run_as_user" to. And if I don't set "run_as_user" it maps root in the pod to root on the host, breaking user-ns isolation.

On the other hand if I use "private:uidmapping=0:296608:65536" to control the mapping myself, I can't set "run_as_user" to  296608 because I get the `cannot use uidmapping or gidmapping if not running as root` error. And if I set "run_as_user" to 0, again I get the root in pod = root in host mapping which I want to avoid.

So K8s users are left without any real options here (as far as I can tell) to generate the basic mapping of root in the pod maps to a non-root Host UID (e.g., 0:296608:65536) on the host.
